### DISCUSSION TITLE
Enforcing Custom Transfer Logic in Custom Coin Modules

### DISCUSSION THREAD

**Discussion by mshakeg** [aptos-labs/aptos-developer-discussions#168] [Questions] [answered]
### Describe your question in detail.

I am currently designing a custom coin module. My goal is to ensure that all transfers of this custom coin type are conducted through a specific transfer function that I will define within the module. This function is intended to incorporate additional logic, such as fees, KYC restrictions, and other necessary features.

The main challenge I'm encountering is enforcing the use of this custom transfer function. Specifically, I want to prevent users from circumventing the custom logic by using standard transfer functions like `aptos_framework::coin::transfer` or `aptos_framework::aptos_account::transfer_coins`. It's crucial that these standard methods do not allow users to bypass the custom logic I'm implementing.

I'm seeking advice on best practices and design patterns within the Move language and the Aptos framework to achieve this. How can I architect my module to ensure that the custom transfer function is the only method used for transactions involving my custom coin?

Any insights, examples, or references to similar implementations would be greatly appreciated.

### What have you tried or looked at?  Or how can we reproduce the error?

I've taken a look at the following from the developer docs, but there doesn't seem to be any mention of what I am trying to achieve

https://aptos.dev/tutorials/your-first-coin/

https://aptos.dev/standards/aptos-coin

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

N/A

**Comment by mshakeg** [comment]
I asked this [on stackoverflow](https://stackoverflow.com/q/78389454/10261711) for further reach

**Comment by BriungRi** [comment]
The current design of `Coin` and `FungibleAsset` doesn't make it possible (as far as I know!) to enforce custom logic during transfer because it enforces a standard for functions like `transfer`. That said, I believe the Aptos Move/Blockchain team is working on a feature called "dynamic dispatch" which may make functionality like you're describing possible.

**Comment by gregnazario** [comment, marked as answer]
Copied from stack overflow.  Keeping in mind that this only applies to Fungible Asset not coin.

Aptos has launched its dispatchable fungible tokens where you can create custom withdraw/transfer functionality overwriting the default functions.

```
module my_addr::my_fungible_asset_example {
// ...
public fun withdraw<T: key>(
    store: Object<T>,
    amount: u64,
    transfer_ref: &TransferRef,
): FungibleAsset {
    // Your custom logic here
}

public fun deposit<T: key>(
    store: Object<T>,
    fa: FungibleAsset,
    transfer_ref: &TransferRef,
) {
    // Your custom logic here
}
// ...}
```
Here's a working example of taxed assets where on every transfer - 10% tax is sent to the deployer account):

https://github.com/aptos-labs/move-by-examples/blob/main/taxed-fungible-asset/move/sources/taxed_fa.move

Read more about the docs: https://aptos.dev/en/build/smart-contracts/fungible-asset#dispatchable-fungible-asset-advanced

AIP (Aptos Improvement Proposal): https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-73.md
